src/git-source-provider.ts

```diff
@@ -0,0 +1,285 @@
+        return [`+refs/heads/${ref}*:refs/remotes/origin/${ref}*`, `+refs/tags/${ref}*:refs/tags/${ref}*`];
+    }
+    // refs/heads/
+    else if (upperRef.startsWith('REFS/HEADS/')) {
+        let branch = ref.substring('refs/heads/'.length);
+        return [`+${ref}:refs/remotes/origin/${branch}`];
+    }
+    // refs/pull/
+    else if (upperRef.startsWith('REFS/PULL/')) {
+        let branch = ref.substring('refs/pull/'.length);
+        return [`+${ref}:refs/remotes/pull/${branch}`];
+    }
+    // refs/tags/
+    else {
+        return [`+${ref}:${ref}`]
+    }
+}
+
+async function getCheckoutInfo(
+    git: gitCommandManager.IGitCommandManager,
+    ref: string,
+    commit: string)
+    : Promise<ICheckoutInfo> {
+
+    if (!ref && !commit) {
+        throw new Error('Args ref and commit cannot both be empty');
+    }
+
+    let result = {} as ICheckoutInfo;
+    let upperRef = (ref || '').toUpperCase();
+
+    // SHA
+    if (!ref) {
+        result.ref = commit;
+    }
+    // refs/heads/
+    if (upperRef.startsWith('REFS/HEADS/')) {
+        let branch = ref.substring('refs/heads/'.length);
+        result.ref = branch;
+        result.startPoint = `refs/remotes/origin/${branch}`;
+    }
+    // refs/pull/
+    else if (upperRef.startsWith('REFS/PULL/')) {
+        let branch = ref.substring('refs/pull/'.length);
+        result.ref = `refs/remotes/pull/${branch}`;
+    }
+    // refs/tags/
+    else if (upperRef.startsWith('REFS/')) {
+        result.ref = ref;
+    }
+    // Unqualified ref, check for a matching branch or tag
+    else {
+        if (await git.branchExists(true, `origin/${ref}`)) {
+            result.ref = ref;
+            result.startPoint = `refs/remotes/origin/${ref}`;
+        }
+        else if (await git.tagExists(`${ref}`)) {
+            result.ref = `refs/tags/${ref}`;
+        }
+        else {
+            throw new Error(`A branch or tag with the name '${ref}' could not be found`);
+        }
+    }
+
+    return result;
+}
+
+async function removeGitConfig(
+    git: gitCommandManager.IGitCommandManager,
+    configKey: string) {
+
+    if (await git.configExists(configKey) &&
+        !(await git.tryConfigUnset(configKey))) {
+
+        // Load the config contents
+        core.warning(`Failed to remove '${configKey}' from the git config. Attempting to remove the config value by editing the file directly.`);
+        let configPath = path.join(git.getWorkingDirectory(), '.git', 'config');
+        fsHelper.fileExistsSync(configPath);
+        let contents = fs.readFileSync(configPath).toString() || '';
+
+        // Filter - only includes lines that do not contain the config key
+        let upperConfigKey = configKey.toUpperCase();
+        let split = contents.split('\n').filter(x => x.toUpperCase().indexOf(upperConfigKey) < 0);
+        contents = split.join('\n');
+
+        // Rewrite the config file
+        fs.writeFileSync(configPath, contents);
+    }
+}
+
+interface ICheckoutInfo {
```

Comment:
put interfaces at the top of the file, or maybe more ideal make it a separate file unless it's part of a class

Reply:
currently its only used privately within this file. i can move to a separate file but i'm not sure that helps readability?

Reply:
rationale being - contained within this module privately is advantageous that it doesnt bloat number of files. it's a tradeoff either way.

Reply:
btw i added eslint and prettier config same as used in the toolkit. should cut down on most style issues

Reply:
i moved this into a separate file along with a couple of the functions when i refactored for unit testing (public now from the helper module, moved to the top)